### ISSUE TITLE
MongoDB Child Date Fields Error - "can't convert from BSON type Object to Date"

### ISSUE THREAD

**Issue by dworrad** [metabase/metabase#2947]
If this is a bug report, please fill in the blanks:
-  I am using the MAC App browser.
-  My computer's OS is OS X El Capitan.
-  I'm running Metabase Locally with MAC app
-  My database is MongoDB v3.2.3
-  My Metabase version is 18.1.1

My Data
```json
{
"_id" : ObjectId("5641f6023ab8180f196eeefc"),
    "email" : "someone@email.com",
    "metadata" : {
        "created" : ISODate("2015-11-11T09:09:30.674Z")
    }
}
```

Using Metadata I try to group this data on metadata.created

I get this error - 
"There was a problem with your question"

The underlying error is - 
"Command failed with error 16006: 'can't convert from BSON type Object to Date' on server localhost:27017....."

I believe that child objects aren't being categorised correctly. i.e The "metadata.created" field is seen as an Object because it's parent is an Object.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by camsaul** [comment]
Related to #1367

**Comment by rajeshbabu999** [comment]
Hi,

I have loaded the data in mongodb in the below format using $date datatype.

{
"_id" : ObjectId("5641f6023ab8180f196eeefc"),
    "email" : "someone@email.com",
    "metadata" : {
        "created" : {$date : 1469561142653}
    }
}

The data is properly loaded as a date value in the database.

I am trying to do some date range queries in metabase.

I am currently getting the below error in mongodb.

**# Command failed with error 16006: 'can't convert from BSON type object to Date' on server** 

are there any updates or fixes for this bug in metabase?

**Comment by FerJmnz** [comment]
Hi,

I have the same problem with nested properties. In my case "**registeredOn**", "**modifiedOn**" and all the date properties.

`"event": {
  "ownerEmail": "test@one.com",
  "device": {
    "typeName": "Automotor",
    "plate": "MEI-1252",
    "chasis": "SP18875",
    "registeredOn": ISODate("2016-09-29T09:11:09.576-05:00"),
    "modifiedOn": ISODate("2016-09-29T09:11:09.576-05:00")
  }
}`

Please help us with any update o fix.

Thanks!!

**Comment by Serfenia** [comment]
Got the same issue when I want to group by a date field that has null values.
Due to the projection stage being before the match stage, you get the issue that it tries to project a NULL field and do calculations with it to get the correct weeks. 

It has big value for us to have this fixed. Please give this a high priority.
Metabase version: 0.23.0

**Comment by farscape2012** [comment]
Got the same issue.
The version is v0.24.1

**Comment by gahabeen** [comment]
Same issue than @Serfenia. 

We're using Timestamp UNIX (seconds) and probably because some values are NULL it won't be able to perform calculation.

**Metabase** version:  0.25.2

This is really important if we want to be using Metabase in production. 

See Question:
http://i.imgur.com/AjH1alk.png

See the full log report:
[metabase-logs-1502612820512.txt](https://github.com/metabase/metabase/files/1220348/metabase-logs-1502612820512.txt)

**Comment by gahabeen** [comment]
@camsaul Here is the reason why queries fails in queries for nested Dates with MongoDB: 

I've analyzed a failing query with the Query Builder and the exact same one made by hand.

Here is a query made by hand which simply looks up if a date equal today's date:
```json
[
  {
    "$project": {
      "time__scheduled__arrival~~~day": {
        "$let": {
          "vars": { "field": "$time.scheduled.arrival" },
          "in": {
            "___date": {
              "$dateToString": { "format": "%Y-%m-%d", "date": "$$field" }
            }
          }
        }
      }
    }
  },
  {
    "$match": {
      "time__scheduled__arrival~~~day": { "$eq": { "___date": "2017-08-15" } }
    }
  },
  { "$group": { "_id": null, "count": { "$sum": 1 } } },
  { "$sort": { "_id": 1 } },
  { "$project": { "_id": false, "count": true } }
]
```

Here is the query builder's one for the same request :
```json
[
  {
    "$project": {
      "time___scheduled___departure~~~day": {
        "$let": {
          "vars": { "field": "$time.scheduled.departure" },
          "in": {
            "___date": {
              "$dateToString": { "format": "%Y-%m-%d", "date": "$$field" }
            }
          }
        }
      },
      "time___scheduled~~~day": {
        "$let": {
          "vars": { "field": "$time.scheduled" },
          "in": {
            "___date": {
              "$dateToString": { "format": "%Y-%m-%d", "date": "$$field" }
            }
          }
        }
      }
    }
  },
  {
    "$match": {
      "time___scheduled___departure~~~day": {
        "$eq": { "___date": "2017-08-15" }
      }
    }
  },
  { "$group": { "_id": "nil", "count": { "$sum": 1 } } },
  { "$sort": { "_id": 1 } },
  { "$project": { "_id": false, "count": true } }
]
```

For some reason the query builder creates the query on the nested targetted field but then also on the parent property.

Not sure how that is generated on your side but it might be a good hint to fix it :)
Hope that helps!

**Comment by cooper667** [comment]
Same issue, assume it's for the reason above

**Comment by B-Stefan** [comment]
Still occurs in version 0.26.2

**Comment by tobiashinz** [comment]
Any plans to fix this?

**Comment by t1maccapp** [comment]
Is there any workaround except changing the actual data?

**Comment by gmaijoe** [comment]
did anyone discover a fix to this?

**Comment by AnushreeDhar** [comment]
try using $unwind on the date ?

**Comment by camsaul** [comment]
It seems like we're generating filter clauses incorrectly for nested columns in MongoDB. Since we have simple steps to repro this issue it should be straightforward to fix

**Comment by LarissaGuder** [comment]
Still occurs in version v0.36.0

**Comment by robdaemon** [comment]
@LarissaGuder can you provide a sample of your data, and the query you're running? I cannot reproduce it given this data (based on an earlier response in this issue:

```
db.nestedData.insertMany([{"event": {
  "ownerEmail": "test@one.com",
  "device": {
    "typeName": "Automotor",
    "plate": "MEI-1252",
    "chasis": "SP18875",
    "dates": {
      "registeredOn": ISODate("2016-09-29T09:11:09.576-05:00"),
      "modifiedOn": ISODate("2016-09-29T09:11:09.576-05:00")
    }
  }
}}, 
{"event": {
  "ownerEmail": "test@example.com",
  "device": {
    "typeName": "Foobar",
    "plate": "ABC123",
    "chasis": "SP18875",
    "dates": {
      "registeredOn": ISODate("2019-09-29T09:11:09.576-05:00"),
      "modifiedOn": ISODate("2019-09-29T09:11:09.576-05:00")
    }
  }
}},
{"event": {
  "ownerEmail": "foo@example.com",
  "device": {
    "typeName": "Foobar",
    "plate": "ABC123",
    "chasis": "SP18875",
    "dates": {
      "registeredOn": null,
      "modifiedOn": ISODate("2020-08-11T09:11:09.576-05:00")
    }
  }
}}]);
```

and a custom query:

![mongo-nested-dates](https://user-images.githubusercontent.com/287213/89948350-cbfe1000-dbda-11ea-94c5-d4f867fb9e37.png)

**Comment by Cheung0927** [comment]
Still occurs in version v0.40.0

**Comment by qnkhuat** [comment]
I can't reproduce this and seems like there are several problems from all the comments
1. Failed to group by a column with ISO date type (from the author of this issue)
2. Failed to group if date column has NULL values (from @ Serfenia)
3. Failed to filter if a column is a text and were configured to cast to Datetime (#18733)

The 1st and 2nd is fixed - tested with current master (b1689921e91e955eb61dbbec9544125dbb1cdef5)

I believe the 3rd is a different problem and should be fixed seperately (check this [comment](https://github.com/metabase/metabase/issues/18733#issuecomment-1112075208)).

If my understand is correct, then this issue could be closed. cc @flamber
